src/qutip_qip/device/processor.py

```diff
@@ -1020,3 +1198,120 @@ def _pulse_interpolate(pulse, tlist):
         pulse.tlist, coeff, kind=kind, bounds_error=False, fill_value=0.0
     )
     return inter(tlist)
+
+
+class Model:
+    """
+    Template class for a physical model representing quantum hardware.
+    The concrete model class does not have to inherit from this,
+    as long as the following methods are defined.
+
+    Parameters
+    ----------
+    num_qubits : int, optional
+        The number of component systems.
+    dims : list, optional
+        The dimension of each component system.
+        Default value is a qubit system of ``dim=[2,2,2,...,2]``.
+    **params :
+        Hardware parameters for the model.
+
+    Attributes
+    ----------
+    num_qubits : int, optional
+        The number of component systems.
+    dims : list, optional
+        The dimension of each component system.
+    params : dict
+        Hardware parameters for the model.
+    """
+
+    def __init__(self, num_qubits, dims=None, **params):
+        self.num_qubits = num_qubits if num_qubits is not None else N
+        self.dims = dims if dims is not None else num_qubits * [2]
+        self.params = deepcopy(params)
+        self._controls = {}
+        self._drift = []
+        self._noise = []
+
+    def get_all_drift(self) -> List[Tuple[Qobj, List[int]]]:
+        """
+        Get all the drift Hamiltonians.
+
+        Returns
+        -------
+        drift_hamiltonian_list : list
+            A list of drift Hamiltonians in the form of
+            ``[(qobj, targets), ...]``.
+        """
+        return self._drift
+
+    def get_control(self, label: Hashable) -> Tuple[Qobj, List[int]]:
+        """
+        Get the control Hamiltonian corresponding to the label.
+
+        Parameters
+        ----------
+        label : hashable object
+            A label that identifies the Hamiltonian.
+
+        Returns
+        -------
+        control_hamiltonian : tuple
+            The control Hamiltonian in the form of ``(qobj, targets)``.
+        """
+        if hasattr(self, "_old_index_label_map"):
+            _old_index_label_map = self._old_index_label_map
+            if isinstance(label, int):
+                label = _old_index_label_map[label]
+        return self._controls[label]
+
+    def get_control_labels(self) -> List[Hashable]:
+        """
+        Get a list of all available control Hamiltonians.
+        Optional, required only when plotting the pulses or
+        using the optimal control algorithm.
+
+        Returns
+        -------
+        label_list : list of hashable objects
+            A list of hashable objects each corresponds to
+            an available control Hamiltonian.
+        """
+        return list(self._controls.keys())
+
+    def get_noise(self) -> List[Noise]:
+        """
+        Get a list of :obj:`.Noise` objects.
+        Single qubit relaxation (T1, T2) are not included here.
```

Comment:
Why are the local T1 and T2 not included here?

Reply:
Yeah there is a trade-off. They are not included here because they are defined just as two numbers and will be handled in the processor. I can add them directly here but then whoever implements their own Model will need to do it themselves.

On the other hand, separating them is indeed a bit strange. Now looking at them again, I think it might be better to include also T1 T2 here. Whoever implements a custom model will need to do their own homework, which should not be that difficult if they are already at this stage.